base_layer/wallet/src/storage/sqlite_db.rs

```diff
@@ -533,6 +403,107 @@ impl WalletBackend for WalletSqliteDatabase {
     }
 }
 
+/// Confirm if database is encrypted or not and if a cipher is provided confirm the cipher is correct.
+/// Unencrypted the database should contain a MasterSecretKey and associated MasterPublicKey
+/// Encrypted the data should contain a Master Public Key in the clear and an encrypted MasterSecretKey
+/// To confirm if the provided Cipher is correct we decrypt the Master PrivateSecretKey and see if it produces the same
+/// Master Public Key that is stored in the db
+fn check_db_encryption_status(
+    database_connection: &WalletDbConnection,
+    cipher: Option<Aes256Gcm>,
+) -> Result<(), WalletStorageError>
+{
+    let conn = database_connection.acquire_lock();
+    let secret_key = WalletSettingSql::get(DbKey::MasterSecretKey.to_string(), &conn)?;
+    let db_public_key = WalletSettingSql::get(DbKey::MasterPublicKey.to_string(), &conn)?;
+
+    if cipher.is_some() && secret_key.is_none() {
+        error!(
+            target: LOG_TARGET,
+            "Cipher is provided but there is no Master Secret Key in DB to decrypt"
+        );
+        return Err(WalletStorageError::InvalidEncryptionCipher);
+    }
+
+    if let Some(sk) = secret_key {
+        let master_secret_key = match CommsSecretKey::from_hex(sk.as_str()) {
+            Ok(sk) => {
+                // This means the key was unencrypted
+                if cipher.is_some() {
+                    error!(
+                        target: LOG_TARGET,
+                        "Cipher is provided but Master Secret Key is not encrypted"
+                    );
+                    return Err(WalletStorageError::InvalidEncryptionCipher);
+                }
+                sk
+            },
+            Err(_) => {
+                // This means the secret key was encrypted. Try decrypt
+                if let Some(cipher_inner) = cipher.clone() {
```

Comment:
nit: 
```suggestion
                if let Some(cipher_inner) = cipher.as_ref() {
```

or pass in a ref `cipher: Option<&Aes256Gcm>` if you really want to save clones